File: cpp/examples/linux/temperature/MCP9808/complete/main.cpp
#include <cstdio>
#include <unistd.h>
#include "I2CConnectionLinux.h"
#include "MCP9808.h"

static volatile uint8_t pendingStatus = 0;
static volatile bool pending = false;

static void onAlert(uint8_t status) {
    pendingStatus = status;
    pending = true;
}

int main() {
    I2CConnectionLinux connection(1, MCP9808Minimal::I2C_ADDRESS);
    MCP9808Full sensor(connection);                         // Create MCP9808 Full driver, (connection)
                                                            // checks MANUFACTURER_ID 0x0054 and DEVICE_ID 0x04

    float t = sensor.readTemperature();                     // Read ambient temperature, () → °C
                                                            // masks TA's 3 status bits, decodes 1/16 °C two's complement
    printf("temperature %.4f C\n", (double)t);

    sensor.setResolution(0.25f);                            // Set resolution, (celsius 0.5|0.25|0.125|0.0625) → bool
                                                            // 0.25 °C step converts in ~65 ms instead of 250 ms
    printf("resolution %.4f C\n", (double)sensor.getResolution());  // Read resolution, () → °C
                                                            // decodes the RESOLUTION register code

    sensor.shutdown();                                      // Enter Shutdown mode, () → void
                                                            // stops conversion; TA keeps its last value
    printf("shutdown %d\n", sensor.isShutdown());              // Check Shutdown mode, () → bool
                                                            // reads CONFIG.SHDN
    sensor.wake();                                          // Leave Shutdown mode, () → void
                                                            // resumes continuous conversion
    usleep(100UL * 1000UL);

    sensor.setUpperLimit(30.0f);                            // Set TUPPER, (celsius °C) → void
                                                            // rounded to the nearest 0.25 °C step
    sensor.setLowerLimit(10.0f);                            // Set TLOWER, (celsius °C) → void
                                                            // rounded to the nearest 0.25 °C step
    sensor.setCriticalLimit(45.0f);                         // Set TCRIT, (celsius °C) → void
                                                            // rounded to the nearest 0.25 °C step
    printf("upper %.2f C\n", (double)sensor.getUpperLimit());  // Read TUPPER, () → °C
                                                            // decodes the 0.25 °C two's-complement boundary
    printf("lower %.2f C\n", (double)sensor.getLowerLimit());  // Read TLOWER, () → °C
                                                            // decodes the 0.25 °C two's-complement boundary
    printf("critical %.2f C\n", (double)sensor.getCriticalLimit());  // Read TCRIT, () → °C
                                                            // decodes the 0.25 °C two's-complement boundary

    sensor.setHysteresis(1.5f);                             // Set hysteresis, (celsius 0|1.5|3.0|6.0) → bool
                                                            // applied on the cooling edge of each boundary only
    printf("hysteresis %.1f C\n", (double)sensor.getHysteresis());  // Read hysteresis, () → °C
                                                            // decodes CONFIG.THYST

    // sensor.lockCriticalLimit();                          // Lock TCRIT, () → void
    //                                                      // irreversible until power-on reset
    // sensor.lockWindowLimits();                           // Lock TUPPER/TLOWER, () → void
    //                                                      // irreversible until power-on reset
    printf("crit locked %d\n", sensor.isCriticalLimitLocked());  // Check TCRIT lock, () → bool
                                                            // reads CONFIG.CRIT_LOCK
    printf("win locked %d\n", sensor.isWindowLimitsLocked());  // Check TUPPER/TLOWER lock, () → bool
                                                            // reads CONFIG.WIN_LOCK

    sensor.configureAlert(MCP9808Full::AlertMode::All,
                          MCP9808Full::AlertOutput::Interrupt,
                          MCP9808Full::AlertPolarity::ActiveLow);  // Configure Alert, (mode=All, output=Comparator, polarity=ActiveLow) → bool
                                                            // sets ALERT_SEL, ALERT_MOD and ALERT_POL together
    sensor.enableAlert();                                   // Enable Alert output, () → void
                                                            // sets CONFIG.ALERT_CNT
    printf("alert asserted %d\n", sensor.isAlertAsserted());  // Check Alert output, () → bool
                                                            // reads the read-only CONFIG.ALERT_STAT

    uint8_t status = sensor.pollInterrupt();                // Read boundary status, () → uint8_t mask
                                                            // TA's live bits: SOURCE_LOWER/UPPER/CRITICAL, nothing cleared
    printf("below lower %d above upper %d critical %d\n",
        (status & MCP9808Full::SOURCE_LOWER) != 0,
        (status & MCP9808Full::SOURCE_UPPER) != 0,
        (status & MCP9808Full::SOURCE_CRITICAL) != 0);
    sensor.clearInterrupt();                                // Clear interrupt-mode Alert, () → void
                                                            // writes CONFIG.INT_CLEAR=1; no effect in comparator mode

    sensor.onInterrupt(onAlert);                            // Subscribe to Alert, (callback, intPin=nullptr) → void
                                                            // callback receives the pollInterrupt() mask; needs an InputPin
    usleep(5000UL * 1000UL);
    if (pending) printf("alert, status mask %u\n", (unsigned)pendingStatus);
    sensor.offInterrupt();                                  // Unsubscribe, () → void
                                                            // detaches the edge handler
    sensor.disableAlert();                                  // Disable Alert output, () → void
                                                            // clears CONFIG.ALERT_CNT
    return 0;
}

File: rust/examples/linux/temperature/mcp9808/complete/src/main.rs
//! Exercises every method in the MCP9808 public API: temperature,
//! resolution, Shutdown mode, the three boundaries and hysteresis, the Alert
//! output, and the boundary-status poll. The one-way lock methods are shown
//! but left commented out — they cannot be undone without a power-on reset.

use linux_embedded_hal::I2cdev;
use periph::chips::temperature::{
    Mcp9808AlertMode, Mcp9808AlertOutput, Mcp9808AlertPolarity, Mcp9808Full, MCP9808_I2C_ADDRESS,
    MCP9808_SOURCE_CRITICAL, MCP9808_SOURCE_LOWER, MCP9808_SOURCE_UPPER,
};
use std::thread::sleep;
use std::time::Duration;

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let mut sensor = Mcp9808Full::new(dev, MCP9808_I2C_ADDRESS).expect("init MCP9808"); // Create MCP9808 Full driver, (i2c, addr=0x18)
                                                                                       // checks MANUFACTURER_ID 0x0054 and DEVICE_ID 0x04

    let t = sensor.read_temperature().expect("read"); // Read ambient temperature, () → f32 °C
                                                      // masks TA's 3 status bits, decodes 1/16 °C two's complement
    println!("temperature {:.4} °C", t);

    sensor.set_resolution(0.25).expect("set_resolution"); // Set resolution, (celsius 0.5|0.25|0.125|0.0625) → ()
                                                          // 0.25 °C step converts in ~65 ms instead of 250 ms
    println!("resolution {} °C", sensor.get_resolution().expect("get_resolution")); // Read resolution, () → f32 °C
                                                                                   // decodes the RESOLUTION register code

    sensor.shutdown().expect("shutdown"); // Enter Shutdown mode, () → ()
                                          // stops conversion; TA keeps its last value
    println!("shutdown {}", sensor.is_shutdown().expect("is_shutdown")); // Check Shutdown mode, () → bool
                                                                        // reads CONFIG.SHDN
    sensor.wake().expect("wake"); // Leave Shutdown mode, () → ()
                                  // resumes continuous conversion
    sleep(Duration::from_millis(100));

    sensor.set_upper_limit(30.0).expect("set_upper_limit"); // Set TUPPER, (celsius °C) → ()
                                                            // rounded to the nearest 0.25 °C step
    sensor.set_lower_limit(10.0).expect("set_lower_limit"); // Set TLOWER, (celsius °C) → ()
                                                            // rounded to the nearest 0.25 °C step
    sensor.set_critical_limit(45.0).expect("set_critical_limit"); // Set TCRIT, (celsius °C) → ()
                                                                  // rounded to the nearest 0.25 °C step
    println!("upper {}", sensor.get_upper_limit().expect("get_upper_limit")); // Read TUPPER, () → f32 °C
                                                                             // decodes the 0.25 °C two's-complement boundary
    println!("lower {}", sensor.get_lower_limit().expect("get_lower_limit")); // Read TLOWER, () → f32 °C
                                                                             // decodes the 0.25 °C two's-complement boundary
    println!("critical {}", sensor.get_critical_limit().expect("get_critical_limit")); // Read TCRIT, () → f32 °C
                                                                                      // decodes the 0.25 °C two's-complement boundary

    sensor.set_hysteresis(1.5).expect("set_hysteresis"); // Set hysteresis, (celsius 0|1.5|3.0|6.0) → ()
                                                         // applied on the cooling edge of each boundary only
    println!("hysteresis {}", sensor.get_hysteresis().expect("get_hysteresis")); // Read hysteresis, () → f32 °C
                                                                                // decodes CONFIG.THYST

    // sensor.lock_critical_limit().expect("lock"); // Lock TCRIT, () → ()
    //                                              // irreversible until power-on reset
    // sensor.lock_window_limits().expect("lock");  // Lock TUPPER/TLOWER, () → ()
    //                                              // irreversible until power-on reset
    println!("crit locked {}", sensor.is_critical_limit_locked().expect("locked")); // Check TCRIT lock, () → bool
                                                                                   // reads CONFIG.CRIT_LOCK
    println!("win locked {}", sensor.is_window_limits_locked().expect("locked")); // Check TUPPER/TLOWER lock, () → bool
                                                                                 // reads CONFIG.WIN_LOCK

    sensor
        .configure_alert(Mcp9808AlertMode::All, Mcp9808AlertOutput::Interrupt, Mcp9808AlertPolarity::ActiveLow)
        .expect("configure_alert"); // Configure Alert, (mode, output, polarity) → ()
                                    // sets ALERT_SEL, ALERT_MOD and ALERT_POL together
    sensor.enable_alert().expect("enable_alert"); // Enable Alert output, () → ()
                                                  // sets CONFIG.ALERT_CNT
    println!("alert asserted {}", sensor.is_alert_asserted().expect("is_alert_asserted")); // Check Alert output, () → bool
                                                                                          // reads the read-only CONFIG.ALERT_STAT

    let status = sensor.poll_interrupt().expect("poll_interrupt"); // Read boundary status, () → u8 mask
                                                                   // TA's live bits: SOURCE_LOWER/UPPER/CRITICAL, nothing cleared
    println!(
        "below lower {} above upper {} critical {}",
        status & MCP9808_SOURCE_LOWER != 0,
        status & MCP9808_SOURCE_UPPER != 0,
        status & MCP9808_SOURCE_CRITICAL != 0
    );
    sensor.clear_interrupt().expect("clear_interrupt"); // Clear interrupt-mode Alert, () → ()
                                                        // writes CONFIG.INT_CLEAR=1; no effect in comparator mode
    sensor.disable_alert().expect("disable_alert"); // Disable Alert output, () → ()
                                                    // clears CONFIG.ALERT_CNT
}
